### ISSUE TITLE
SetResponseHeaderLayer et al are hard to use with literal HeaderValues

### ISSUE THREAD

**Issue by Roguelazer** [tower-rs/tower-http#127]
### Version

0.1.1

### Platform
Linux

### Description

SetResponseHeaderLayer (and associated functions) use the MakeHeaderValue trait for the value. This is implemented on `HeaderValue` itself (for setting a constant) but I'm at a loss for how to use it because of the unrestricted type `T` in its generic signature

Given this Axum code:
```rust
    let hostname = HeaderValue::from_static("example");

    let app = Router::new()
        .route("/v1/ok", get(handle_ok))
        .layer(tower_http::set_header::SetResponseHeaderLayer::overriding(
            "x-hostname".parse().unwrap(),
            hostname,
        ));
```

Compilation fails with:

```
error[E0282]: type annotations needed
   --> src/main.rs:198:16
    |
198 |         .layer(tower_http::set_header::SetResponseHeaderLayer::overriding(
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T`
```

I have no idea how to pass this type. It's not constrained in the docs and I don't even know what it //is//.

**Comment by davidpdrsn** [comment]
While I wish there was a good solution for this I don't think there is.

The docs have an example https://docs.rs/tower-http/0.1.1/tower_http/set_header/response/index.html#example

In your particular case this should do it:

```rust
.layer(tower_http::set_header::SetResponseHeaderLayer::<_, axum::body::Body>::overriding(
    "x-hostname".parse().unwrap(),
    axum::http::HeaderValue::from_static("example"),
))
```

**Comment by Roguelazer** [comment]
I almost think it might make sense to just have a non-generic version of the Layer that takes a constant?

**Comment by davidpdrsn** [comment]
> I almost think it might make sense to just have a non-generic version of the Layer that takes a constant?

I see where you're coming from but that would require duplicating the entire implementation. I don't think that's worth it.

If it's really something you want you can always write your own middleware. There is an example in axum's docs you use https://docs.rs/axum/0.2.0/axum/#writing-your-own-middleware

**Comment by jplatte** [comment]
> I see where you're coming from but that would require duplicating the entire implementation. I don't think that's worth it.

I am certain that this is not true, and the changes required to make it work aren't even that crazy. I'm about to open a PR to improve `set_header` a bit that might make you scratch your head about how things worked *previously*, and hope to have an actual fix for this ready soon.

**Comment by davidpdrsn** [comment]
Haha that sounds cool :P

**Comment by jplatte** [comment]
Fun fact: This can also be worked around by setting the second generic parameter to *any other type* that's not `axum::body::Body`.

**Comment by jplatte** [comment]
Not sure whether it makes sense to reopen this, but there is now a proper fix for this: #148 (unfortunately a breaking change, but there is nothing that can be done about that).
